e2etests/test/module_methods.js

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+'use strict';
+
+var ModuleIdentityHelper = require('./module_identity_helper.js');
+var assert = require('chai').assert;
+var debug = require('debug')('e2etests:module-methods');
+var Amqp = require('azure-iot-device-amqp').Amqp;
+var AmqpWs = require('azure-iot-device-amqp').AmqpWs;
+//var Mqtt = require('azure-iot-device-mqtt').Mqtt;
+//var MqttWs = require('azure-iot-device-mqtt').MqttWs;
+
+var transportsToTest = [ Amqp, AmqpWs ];
+
+describe('module methods', function() {
+  this.timeout(46000);
+
+  transportsToTest.forEach(function(Transport) {
+    describe('using ' + Transport.name, function() {
+      var testModule = {};
+
+      before(function(done) {
+        ModuleIdentityHelper.createModule(testModule, Transport, done);
+      });
+
+      after(function(done) {
+        ModuleIdentityHelper.cleanUpAfterTest(testModule, done);
+      });
+
+      it ('can receive a method call', function(done) {
+        var methodName = 'my_method';
+        var requestPayload = {
+          fakePayloadKey: '__FAKE_PAYLOAD_VALUE__'
+        };
+        var methodResult = 400;
+        var responsePayload = {
+          anotherFakePayloadKey: '__ANOTHER_FAKE_PAYLOAD_VALUE__'
+        };
+        var methodParams = {
+          methodName: methodName,
+          payload: requestPayload,
+          responseTimeoutInSeconds: 15
+        };
+
+        testModule.deviceClient.onDeviceMethod(methodName, function(request, response) {
+          debug('received method call');
+          assert.strictEqual(request.methodName, methodName);
+          assert.deepEqual(request.payload, requestPayload);
+
+          debug('sending method response with statusCode: ' + methodResult);
+          response.send(methodResult, responsePayload, function(err) {
+            debug('response.send returned ' + (err ? err : 'success'));
+            assert(!err);
+          });
+        });
+
+        // Waiting for an arbitrary 2 seconds because we don't know when all the links above have been established.
```

Comment:
>Waiting for an arbitrary 2 seconds [](start = 11, length = 34)

we need retries for this instead. if the service client doesn't have a PBI for invokeMethod retries, let's add one.

Reply:
Added 2366350

---
In reply to: [180597799](https://github.com/Azure/azure-iot-sdk-node/pull/271#discussion_r180597799) [](ancestors = 180597799)